k2/csrc/fsa_test.cc

```diff
@@ -0,0 +1,316 @@
+      3 4 -1
+      4
+  )";
+  auto fsa = StringToFsa(s);
+  Cfsa cfsa(*fsa);
+
+  EXPECT_EQ(cfsa.num_states, 5);
+  EXPECT_EQ(cfsa.begin_arc, 0);
+  EXPECT_EQ(cfsa.end_arc, 5);
+  EXPECT_EQ(cfsa.arc_indexes, fsa->arc_indexes.data());
+  EXPECT_EQ(cfsa.arcs, fsa->arcs.data());
+
+  EXPECT_EQ(cfsa.NumStates(), 5);
+  EXPECT_EQ(cfsa.FinalState(), 4);
+}
+
+TEST(Cfsa, ConstructorEmptyFsa) {
+  Fsa fsa;
+  Cfsa cfsa(fsa);
+  EXPECT_EQ(cfsa.num_states, 0);
+  EXPECT_EQ(cfsa.begin_arc, 0);
+  EXPECT_EQ(cfsa.end_arc, 0);
+  EXPECT_EQ(cfsa.arc_indexes, nullptr);
+  EXPECT_EQ(cfsa.arcs, nullptr);
+}
+
+TEST(GetCfsaVecSize, Empty) {
+  Cfsa cfsa;
+  size_t bytes = GetCfsaVecSize(cfsa);
+  // 20-byte header             (20)
+  // 44-byte padding            (64)
+  // 8-byte state_offsets_array (72)
+  // 56-byte padding            (128)
+  // 4-byte arc_indexes_array   (132)
+  EXPECT_EQ(bytes, 132u);
+
+  std::vector<Cfsa> cfsa_vec;
+  cfsa_vec.push_back(cfsa);
+  bytes = GetCfsaVecSize(cfsa_vec);
+  EXPECT_EQ(bytes, 132u);
+}
+
+TEST(GetCfsaVecSize, NonEmpty) {
+  std::string s = R"(
+      0 1 1
+      0 2 2
+      1 3 3
+      2 3 3
+      3 16 -1
+      16
+  )";
+  auto fsa = StringToFsa(s);
+  Cfsa cfsa(*fsa);
+
+  size_t bytes = GetCfsaVecSize(cfsa);
+  // 20-byte header             (20)
+  // 44-byte padding            (64)
+  // 8-byte state_offset_array  (72)
+  // 56-byte padding            (128)
+  // 72-byte arc_indexes_array  (200)
+  // 4-byte padding             (204) -> to be multiple of sizeof(Arc)
+  // 60-byte arcs_array         (264)
+  EXPECT_EQ(bytes, 264u);
+  // Note that there are 5 arcs, sizeof(Arc) == 12.
+  // There are 17 states and each state needs 4 bytes
+  // and the last state is repeated, so the arc_indexes
+  // array needs 18*4 = 72-byte
+
+  {
+    std::vector<Cfsa> cfsa_vec;
+    cfsa_vec.push_back(cfsa);
+    bytes = GetCfsaVecSize(cfsa);
+    EXPECT_EQ(bytes, 264u);
+  }
+}
+
+TEST(GetCfsaVecSize, NonEmptyMutlipeFsas) {
+  std::string s1 = R"(
+      0 1 1
+      0 2 2
+      1 3 3
+      2 3 3
+      3 16 -1
+      16
+  )";
+  auto fsa1 = StringToFsa(s1);  // 5 arcs, 17 states
+  Cfsa cfsa1(*fsa1);
+
+  std::string s2 = R"(
+      0 1 1
+      0 2 2
+      1 3 3
+      3 10 -1
+      10
+  )";
+  auto fsa2 = StringToFsa(s2);  // 4 arcs, 11 states
+  Cfsa cfsa2(*fsa2);
+
+  std::vector<Cfsa> cfsa_vec = {cfsa1, cfsa2};
+
+  size_t bytes = GetCfsaVecSize(cfsa_vec);
+  // 28 states,9 arcs
+  //
+  // 20-byte header             (20)
+  // 44-byte padding            (64)
+  // 12-byte state_offset_array (76)
+  // 52-byte padding            (128)
+  // 120-byte arc_indexes_array (248)
+  // 4-byte padding             (252) -> to be multiple of sizeof(Arc)
+  // 108-byte arcs_array        (360)
+  EXPECT_EQ(bytes, 360u);
+}
+
+TEST(CfsaVec, Empty) {
+  Cfsa cfsa;
+  std::vector<Cfsa> cfsas;
+  size_t bytes = GetCfsaVecSize(cfsas);
+  std::unique_ptr<void, decltype(&MemFree)> data(MemAlign(bytes, 64), &MemFree);
```

Comment:
why do we need align here?  I just thought we have no requirement on the end of data (to be aligned).

Reply:
There is no real requirement for alignment.  It's just a nice-to-have as in some cases it might improve efficiency.  For test code it's probably not necessary, but also doesn't hurt.  We can assume that when we allocate tensors from things like PyTorch, it will be memory aligned to whatever boundaries the hardware cares about.

Reply:
OK, then I suggest we add what you said as comments in the test code here if we want to keep the alignment code (so the readers will not misunderstand when reading the test code here)

Reply:
Not sure what will happen if the returned pointer is not aligned to `alignof(CfsaVecHeader)`
and `alignof(Arc)` when `MemAlign` is not used.